### ISSUE TITLE
Unnecessary CPU usage because TST falsely thinks tab is loading and shows tab loading animation

### ISSUE THREAD

**Issue by azizLIGHT** [piroor/treestyletab#2275]
## Short description
Using addons like LoadTabOnSelect that prevent loading of tabs until tab is selected causes TST tab loading animation ("throbber") to animate forever on tabs that should be considered loaded. The animation causes undue CPU usage. The effect is amplified when multiple tabs are simultaneously loaded using "open all in tabs" right click context menu option on a bookmarks folder containing many bookmarks

## Steps to reproduce

 1. Start Firefox Nightly with clean profile.
 2. Install TST 
 3. Install [LoadTabOnSelect](https://addons.mozilla.org/en-US/firefox/addon/loadtabonselect/?src=search) OR [LoadTabOnSelect Evolution](https://addons.mozilla.org/en-US/firefox/addon/loadtabonselect-evolution/?src=search) (also works on [Load Background Tabs Lazily](https://addons.mozilla.org/en-US/firefox/addon/load-background-tabs-lazily/?src=search))
 4. Import Bookmarks from [HTML](https://gist.githubusercontent.com/azizLIGHT/585bb279788a26b5dda4f24710da0172/raw/df49399400882bd494438f280ca2d367be770b71/bookmarks.html) using Ctrl+Shift+B -> Import Bookmarks from HTML
5. Open 200 tabs of google.com using right click on bookmarks folder google -> "Open all Tabs" 

## Expected result
Since Firefox tab strip at top has no throbber animation, and tab is considered loaded, for all intents and purposes, tabs in TST should be shown as completely loaded and have no throbber animation
![image](https://user-images.githubusercontent.com/8013867/57577955-329df680-748b-11e9-8849-88a0c5baa7c2.png)

## Actual result
Some tabs are considered not loaded fully by TST and show throbber animation forever
![nightly-TSTdev-throbber](https://user-images.githubusercontent.com/8013867/57577977-9d4f3200-748b-11e9-80da-d2f2bef5b25f.gif)

## Environment

 * Platform (OS): Windows 7
 * Version of Firefox: Nightly 68.0a1 (2019-05-11) (64-bit)
 * Version (or revision) of Tree Style Tab: 3.0.11.9109
 * Version of LoadTabOnSelect: 1.2

## Extra details
* Closing throbbing tab takes a long time. 
* Expanding/collapsing group containing throbbing tab also takes a *really* long time.
* TST group tab containing throbbing tabs is also throbbing forever itself
* Opening Bookmarks -> Firefox Nightly Resources -> "Open All in tabs" achieves same effect sometimes. Note: the "Crash Reports" bookmark with url about:crashes throbs forever in TST even after clicking and loading the tab unlike the other tabs.
![nightly-TSTdev-throbber2](https://user-images.githubusercontent.com/8013867/57578042-7abe1880-748d-11e9-8b92-3e5a1ef0560d.gif)
* Small detail: throbbing animation is not in sync

**Comment by piroor** [comment]
I've investigated a little to know what's happen when 200 tabs are opened.

* 200 tabs are become "completed" immediately in the background page.
* Only some tabs are detected as "completed" in the sidebar, and rest tabs are treated as "loading" and become "completed" step by step, on my environment.
  * In other words, on my environment the problem "Some tabs are considered not loaded fully by TST and show throbber animation forever" didn't happen.
* Closing and reopening the sidebar fetch latest status of tabs from the background page, so all tabs in the sidebar become "completed" after reopening.

For consistency TST's sidebar processes events around tabs step by step. Otherwise you may see broken order, broken appearance, and other breakages around tree of tabs produced from asynchronously notified tab events. But as the result step-by-step processing of buffered events is too slow on such cases, and I need to do some more optimizations.

**Comment by piroor** [comment]
Finally I've realized that updates notified between "tab is creating" and "tab is created" were unexpectedly ignored. What happened on a problematic case like you reported is:

1. Firefox notifies a message to `tabs.onCreated` listeners.
2. TST receives the message and starts to initialize a tab object with `"status" = "loading"`.
3. Firefox notifies a message to `webRequest.onBeforeRequest` listeners.
4. Another addon receives the message and cancels it.
5. Firefox notifies a message `"status" ="complete"` to `tabs.onUpdated` listeners.
6. TST receives the message, but the corresponding tab object is not initialized yet, so the update is not applied.
7. TST finishes initialization of the tab object.

So I've introduced a change with the commit 1385fab to apply the `"status" ="complete"` notification (at the step 5 above) certainly, and this problem should be fixed.
